src/drivers/intel/cavs/mn.c

```diff
@@ -142,73 +171,190 @@ static bool find_mn(uint32_t freq, uint32_t bclk,
+ *	  Clock that can use just SCR is preferred.
+ *	  M/N other than 1/1 is used only if there are no other possibilities.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] scr_div SCR divisor.
+ * \param[out] m M value of M/N divider.
+ * \param[out] n N value of M/N divider.
+ * \return index of suitable clock if could find it, -1 otherwise.
+ */
+static int find_bclk_source(uint32_t bclk,
+			    uint32_t *scr_div, uint32_t *m, uint32_t *n)
 {
-	uint32_t i2s_m = 1;
-	uint32_t i2s_n = 1;
-	uint32_t mdivc;
 	int i;
-	int clk_index = -1;
-	int ret = 0;
 
-	*out_need_ecs = false;
+	/* searching the smallest possible bclk source */
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++)
+		if (ssp_freq[i].freq % bclk == 0) {
+			*scr_div = ssp_freq[i].freq / bclk;
+			return i;
+		}
 
-	spin_lock(mn_lock);
-	mdivc = mn_reg_read(0x0);
+	/* check if we can get target BCLK with M/N */
+	for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++)
+		if (find_mn(ssp_freq[i].freq, bclk,
+			    scr_div, m, n))
+			return i;
 
-	/* searching the smallest possible bclk source */
-	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
-		if (bclk_rate > ssp_freq[i].freq)
-			break;
+	return -1;
+}
 
-		if (ssp_freq[i].freq % bclk_rate == 0)
-			clk_index = i;
+/**
+ * \brief Checks if given clock is used as source for any BCLK.
+ * \param[in] clk_src Bit clock source.
+ * \return true if any port use given clock source, false otherwise.
+ */
+static inline bool is_bclk_source_in_use(enum bclk_source clk_src)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(bclk_sources); i++)
+		if (bclk_sources[i] == clk_src)
+			return true;
+	return false;
+}
+
+/**
+ * \brief Configures M/N source clock for BCLK.
+ *	  All ports that use M/N share the same source, so it should be changed
+ *	  only if there are no other ports using M/N already.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] scr_div SCR divisor.
+ * \param[out] m M value of M/N divider.
+ * \param[out] n N value of M/N divider.
+ * \return 0 on success, error code otherwise.
+ */
+static inline int setup_initial_bclk_mn_source(uint32_t bclk, uint32_t *scr_div,
+					       uint32_t *m, uint32_t *n)
+{
+	int clk_index = find_bclk_source(bclk, scr_div, m, n);
+
+	if (clk_index < 0) {
+		trace_mn_error("error: BCLK %d, no valid source", bclk);
+		return -EINVAL;
 	}
 
-	if (clk_index >= 0) {
-		mdivc |= MNDSS(ssp_freq_sources[clk_index]);
-		*out_scr_div = ssp_freq[clk_index].freq / bclk_rate;
-
-		/* select M/N output for bclk in case of Audio Cardinal
-		 * or PLL Fixed clock.
-		 */
-		if (ssp_freq_sources[clk_index] != SSP_CLOCK_XTAL_OSCILLATOR)
-			*out_need_ecs = true;
-	} else {
-		/* check if we can get target BCLK with M/N */
-		for (i = 0; i <= MAX_SSP_FREQ_INDEX; i++) {
-			if (find_mn(ssp_freq[i].freq, bclk_rate,
-				    out_scr_div, &i2s_m, &i2s_n)) {
-				clk_index = i;
-				break;
-			}
-		}
+	bclk_source_mn_clock = clk_index;
+
+	mn_reg_write(0x0, (mn_reg_read(0x0) |
```

Comment:
@jajanusz can you please add a comment here as to what reg we're writing to here?

Reply:
Yea, all numbers like thta just should be symbols, just forgot about that in haste ;)

Reply:
Changed to defines